### PR TITLE
Update to generate v2 snapshot from v3 state

### PR DESCRIPTION
**Pull request by geetasg**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.
Related to https://github.com/etcd-io/etcd/issues/12913

e2e test is TestDowngradeUpgradeClusterOf3WithSnapshot

### CODE REVIEW

**Comment by ahrtr** [comment]
Please take a look at the e2e failures.

e.g.,
`FAIL: TestV2DeprecationSnapshotMatches`

**Comment by geetasg** [comment]
investigating test failure.

**Comment by ahrtr** [comment]
Overall looks good to me, with just a minor comment.

**Comment by geetasg** [comment]
updated as per review feedback

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you!

File: tests/e2e/v2store_deprecation_test.go [hunk 1]
```diff
@@ -125,15 +129,12 @@ func TestV2DeprecationSnapshotMatches(t *testing.T) {
 	assert.NoError(t, epc.Close())
 
 	assertSnapshotsMatch(t, oldMemberDataDir, newMemberDataDir, func(data []byte) []byte {
-		// Patch cluster version
```
Comment (serathius) [inline tests/e2e/v2store_deprecation_test.go:128]: Why remove this?

File: tests/e2e/v2store_deprecation_test.go [hunk 2]
```diff
@@ -125,15 +129,12 @@ func TestV2DeprecationSnapshotMatches(t *testing.T) {
...
-		data = bytes.Replace(data, []byte("3.5.0"), []byte("X.X.X"), -1)
-		data = bytes.Replace(data, []byte("3.6.0"), []byte("X.X.X"), -1)
 		// Patch members ids
 		for i, mid := range members1 {
-			data = bytes.Replace(data, []byte(fmt.Sprintf("%x", mid)), []byte(fmt.Sprintf("member%d", i+1)), -1)
```
Comment (serathius) [inline tests/e2e/v2store_deprecation_test.go:133]: Why change this?

File: tests/e2e/v2store_deprecation_test.go [hunk 3]
```diff
@@ -250,7 +251,23 @@ func assertSnapshotsMatch(t testing.TB, firstDataDir, secondDataDir string, patc
 		if err != nil {
 			t.Fatal(err)
 		}
-		assert.Equal(t, openSnap(patch(firstSnapshot.Data)), openSnap(patch(secondSnapshot.Data)))
+		assertMembershipEqual(t, openSnap(patch(firstSnapshot.Data)), openSnap(patch(secondSnapshot.Data)))
```
Comment (serathius) [inline tests/e2e/v2store_deprecation_test.go:254]: Why change this? Doesn't seem related changing the logic, looks like additional debug info? Could you split unrelated changes to PR?

File: server/etcdserver/server.go [hunk 4]
```diff
@@ -2067,7 +2067,6 @@ func (s *EtcdServer) applyConfChange(cc raftpb.ConfChange, confState *raftpb.Con
 
 // TODO: non-blocking snapshot
 func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
-	clone := s.v2store.Clone()
```
[the remark this answers is not in the document]
Comment (ahrtr) [reply server/etcdserver/server.go:2070]: Good catch! You may want to call `GetMembershipInfoInV2Format` outside the goroutine (right after the `s.KV().Commit()`)

[the remark this answers is not in the document]
Comment (serathius) [reply server/etcdserver/server.go:2070]: Yea, it's not like I have made the same bug when I was implementing it myself.

File: tests/e2e/v2store_deprecation_test.go (hunk 1 above)
Comment (geetasg) [reply tests/e2e/v2store_deprecation_test.go:128]: please refer to the comment above - https://github.com/etcd-io/etcd/pull/16418#issuecomment-1679344123.

File: tests/e2e/v2store_deprecation_test.go (hunk 2 above)
Comment (geetasg) [reply tests/e2e/v2store_deprecation_test.go:133]: please refer to the comment above - https://github.com/etcd-io/etcd/pull/16418#issuecomment-1679344123.

File: server/etcdserver/server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (geetasg) [reply server/etcdserver/server.go:2070]: updated to fix this.

File: server/etcdserver/server_test.go [hunk 5]
```diff
@@ -1098,7 +1088,8 @@ func TestSnapshotOrdering(t *testing.T) {
 	n := newNopReadyNode()
 	st := v2store.New()
```
[the remark this answers is not in the document]
Comment (geetasg) [reply server/etcdserver/server_test.go:1089]: Currently used by v2 applied in the test. I will remove this when we get to removing the v2 applier. If removed now, the test panics in 
```
        lg.Info("restoring v2 store")
        if err := s.v2store.Recovery(toApply.snapshot.Data); err != nil {
                lg.Panic("failed to restore v2 store", zap.Error(err))
        }

stack
	/root/go/pkg/mod/go.uber.org/zap@v1.25.0/zapcore/entry.go:196 +0x65
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00035a270, {0xc0004cea80, 0x2, 0x2})
	/root/go/pkg/mod/go.uber.org/zap@v1.25.0/zapcore/entry.go:262 +0x3ec
go.uber.org/zap.(*Logger).Panic(0xe7d89b?, {0xe89e77?, 0xd63f00?}, {0xc0004cea80, 0x2, 0x2})
	/root/go/pkg/mod/go.uber.org/zap@v1.25.0/logger.go:258 +0x59
go.etcd.io/etcd/pkg/v3/schedule.(*fifo).executeJob.func1()
	/home/ec2-user/geetasg/etcd/pkg/schedule/schedule.go:202 +0x2cb
panic({0xd63f00, 0x15b5a00})
	/usr/local/go/src/runtime/panic.go:884 +0x213
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).applySnapshot(0xc000368000, 0xc0004ce780, 0xc000546600)
	/home/ec2-user/geetasg/etcd/server/etcdserver/server.go:1050 +0x1167
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).applyAll(0xc000368000, 0xc0004ce780, 0xc000546600)
	/home/ec2-user/geetasg/etcd/server/etcdserver/server.go:923 +0x46
go.etcd.io/etcd/server/v3/etcdserver.(*EtcdServer).run.func8({0xc0004f7fa0, 0xc0003af320})
	/home/ec2-user/geetasg/etcd/server/etcdserver/server.go:847 +0x25
go.etcd.io/etcd/pkg/v3/schedule.job.Do(...)
	/home/ec2-user/geetasg/etcd/pkg/schedule/schedule.go:41
go.etcd.io/etcd/pkg/v3/schedule.(*fifo).executeJob(0xc0004f7f90?, {0xfcddb8?, 0xc00021b2c0?}, 0x0?)
	/home/ec2-user/geetasg/etcd/pkg/schedule/schedule.go:206 +0x88
go.etcd.io/etcd/pkg/v3/schedule.(*fifo).run(0xc0001e4af0)
	/home/ec2-user/geetasg/etcd/pkg/schedule/schedule.go:187 +0x11a
created by go.etcd.io/etcd/pkg/v3/schedule.NewFIFOScheduler
	/home/ec2-user/geetasg/etcd/pkg/schedule/schedule.go:101 +0x18b

```

File: tests/e2e/v2store_deprecation_test.go (hunk 3 above)
Comment (serathius) [reply tests/e2e/v2store_deprecation_test.go:254]: > The events on v2store and indices on v2store wont match between the two snapshots anymore

Not sure I understand. It's not clear from code changes. Please separate a PR that updates the test. PR should be organized in a way to clearly show what is changing, if we need to change validation, let's do that separately from business logic.

Comment (geetasg) [reply tests/e2e/v2store_deprecation_test.go]: Adding more clarification on the test change -
Original test was checking that the v2 snapshot content created by older and newer version matched wrt indices on the v2store and events on the v2 store. With the new logic, the indices and events on v2store will not match. As a result we saw a test failure. The test is being updated to verify membership match instead. Please let me know if you have more questions. I would like to confirm that we are on same page wrt this change in the test.

Created separate PR for test changes as per review feedback above (https://github.com/etcd-io/etcd/pull/16441) . The test change will need to  be merged before this PR can pass. Converting this one to draft and removing the test change. The test will start failing until the test PR merges.

Comment (ahrtr) [reply tests/e2e/v2store_deprecation_test.go]: Do not understand why we have to create a separate PR to change the validation. We only care about the membership data in the snapshot, it's reasonable to only compare the membership data.

Anyway, approved the other PR https://github.com/etcd-io/etcd/pull/16441.

**Comment by ahrtr** [comment]
Please squash the commits (Note: rebase PR instead of merging main into dev branch next time).

**Comment by geetasg** [comment]
updated commits

File: server/etcdserver/server.go [hunk 6]
```diff
@@ -2075,16 +2074,12 @@ func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
 	// So KV().Commit() cannot run in parallel with toApply. It has to be called outside
 	// the go routine created below.
 	s.KV().Commit()
+	d := GetMembershipInfoInV2Format(s.Logger(), s.cluster)
```
Comment (serathius) [inline server/etcdserver/server.go:2077]: Overall note about code changes. Please avoid unnecessary reordering between functions in PR, it adds additional review strain. Like in here between `s.KV().Commit()`, `GetMembershipInfoInV2Format(s.Logger(), s.cluster)`. I don't think it matters here, but takes a minute or two to think about this.

**Comment by serathius** [comment]
LGTM, but I would like to double check something. This change mostly depends on `TestDowngradeUpgradeClusterOf*` for testing. We downgrade the cluster and expect it to work with and without snapshot. 

However I noticed that we are not testing contents of membership correctly. We confirm that members before are equal members after, however we don't really do any member changes. So there is no possibility that would allow  membership to differ.

**Comment by serathius** [comment]
PTAL https://github.com/etcd-io/etcd/pull/16457

**Comment by serathius** [comment]
Validated that this PR works with https://github.com/etcd-io/etcd/pull/16457, so we can proceed.
Still, please add your suggestions about more testing for V3 -> V2 snapshot generation. We need to be 100% sure this is correct, and the only way to do that is proper e2e testing.

**Review by serathius — APPROVED** [review]

File: server/etcdserver/server.go (hunk 6 above)
Comment (geetasg) [reply server/etcdserver/server.go:2077]: Created https://github.com/etcd-io/etcd/pull/16460 to keep the order same to simplify reasoning / avoid the unnecessary strain.

File: tests/e2e/v2store_deprecation_test.go (hunk 3 above)
Comment (geetasg) [reply tests/e2e/v2store_deprecation_test.go]: Thanks both. Separate CR was used for test code as per review feedback.
